internal/api/mfa.go

```diff
@@ -388,6 +485,104 @@ func (a *API) challengeTOTPFactor(w http.ResponseWriter, r *http.Request) error
 	})
 }
 
+func (a *API) challengeWebAuthnFactor(w http.ResponseWriter, r *http.Request) error {
+	ctx := r.Context()
+	db := a.db.WithContext(ctx)
+	config := a.config
+
+	user := getUser(ctx)
+	factor := getFactor(ctx)
+	ipAddress := utilities.GetIPAddress(r)
+
+	params := &ChallengeFactorParams{}
+	if err := retrieveRequestParams(r, params); err != nil {
+		return err
+	}
+	if params.WebAuthn == nil {
+		return badRequestError(ErrorCodeValidationFailed, "WebAuthn config required")
+	}
+	webAuthn, err := params.WebAuthn.ToConfig()
+	if err != nil {
+		return err
+	}
+	var response *ChallengeFactorResponse
+	var ws *models.WebAuthnSessionData
+	var challenge *models.Challenge
+	if factor.IsUnverified() {
+		options, session, err := webAuthn.BeginRegistration(user)
+		if err != nil {
+			return internalServerError("error generating WebAuthn registration data").WithInternalError(err)
+		}
+		ws = &models.WebAuthnSessionData{
+			SessionData: session,
+		}
+		challenge = ws.ToChallenge(factor.ID, ipAddress)
+
+		response = &ChallengeFactorResponse{
+			CredentialCreationOptions: options,
+			Type:                      factor.FactorType,
+			ID:                        challenge.ID,
+		}
+
+	} else if factor.IsVerified() {
+		options, session, err := webAuthn.BeginLogin(user)
+		if err != nil {
+			return err
+		}
+		ws = &models.WebAuthnSessionData{
+			SessionData: session,
+		}
+		challenge = ws.ToChallenge(factor.ID, ipAddress)
+		response = &ChallengeFactorResponse{
+			CredentialRequestOptions: options,
+			Type:                     factor.FactorType,
+			ID:                       challenge.ID,
+		}
+
+	}
+
+	err = db.Transaction(func(tx *storage.Connection) error {
+		if terr := factor.WriteChallengeToDatabase(tx, challenge); terr != nil {
+			return terr
+		}
+		return nil
+
+	})
+	if err != nil {
+		return err
+	}
+	response.ExpiresAt = challenge.GetExpiryTime(config.MFA.ChallengeExpiryDuration).Unix()
+
+	return sendJSON(w, http.StatusOK, response)
+
+}
+
+func (a *API) validateChallenge(r *http.Request, db *storage.Connection, factor *models.Factor, challengeID uuid.UUID) (*models.Challenge, error) {
+	config := a.config
+	currentIP := utilities.GetIPAddress(r)
+
+	challenge, err := factor.FindChallengeByID(db, challengeID)
+	if err != nil {
+		if models.IsNotFoundError(err) {
+			return nil, notFoundError(ErrorCodeMFAFactorNotFound, "MFA factor with the provided challenge ID not found")
```

Comment:
Generally HTTP 404 should be reserved when the URL doesn't map to a resource. Since the challenge ID comes from the request body, this looks more like an `unprocessableEntityError` or potentially `badRequestError`.